subxt/examples/unstable_light_client_parachains.rs

```diff
@@ -0,0 +1,101 @@
+use std::{iter, num::NonZeroU32};
+use subxt::{
+    client::{LightClient, RawLightClient},
+    PolkadotConfig,
+};
+
+// Generate an interface that we can use from the node's metadata.
+#[subxt::subxt(runtime_metadata_path = "../artifacts/polkadot_metadata_small.scale")]
+pub mod polkadot {}
+
+const POLKADOT_SPEC: &str = include_str!("../../artifacts/demo_chain_specs/polkadot.json");
+const ASSET_HUB_SPEC: &str =
+    include_str!("../../artifacts/demo_chain_specs/polkadot_asset_hub.json");
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // The smoldot logs are informative:
+    tracing_subscriber::fmt::init();
+
+    // Connecting to a parachain is a multi step process.
+
+    // Step 1. Construct a new smoldot client.
+    let mut client = subxt_lightclient::Client::new(subxt_lightclient::DefaultPlatform::new(
+        "subxt-example-light-client".into(),
+        "version-0".into(),
+    ));
+
+    // Step 2. Connect to the relay chain of the parachain. For this example, the Polkadot relay chain.
+    let polkadot_connection = client
+        .add_chain(subxt_lightclient::AddChainConfig {
+            specification: POLKADOT_SPEC,
+            json_rpc: subxt_lightclient::AddChainConfigJsonRpc::Enabled {
+                max_pending_requests: NonZeroU32::new(128).unwrap(),
+                max_subscriptions: 1024,
+            },
+            potential_relay_chains: iter::empty(),
+            database_content: "",
+            user_data: (),
+        })
+        .expect("Light client chain added with valid spec; qed");
+    let polkadot_json_rpc_responses = polkadot_connection
+        .json_rpc_responses
+        .expect("Light client configured with json rpc enabled; qed");
+    let polkadot_chain_id = polkadot_connection.chain_id;
+
+    // Step 3. Connect to the parachain. For this example, the Asset hub parachain.
+    let assethub_connection = client
+        .add_chain(subxt_lightclient::AddChainConfig {
+            specification: ASSET_HUB_SPEC,
+            json_rpc: subxt_lightclient::AddChainConfigJsonRpc::Enabled {
+                max_pending_requests: NonZeroU32::new(128).unwrap(),
+                max_subscriptions: 1024,
+            },
+            // The chain specification of the asset hub parachain mentions that the identifier
+            // of its relay chain is `polkadot`.
+            potential_relay_chains: [polkadot_chain_id].into_iter(),
+            database_content: "",
+            user_data: (),
+        })
+        .expect("Light client chain added with valid spec; qed");
+    let parachain_json_rpc_responses = assethub_connection
+        .json_rpc_responses
+        .expect("Light client configured with json rpc enabled; qed");
+    let parachain_chain_id = assethub_connection.chain_id;
+
+    // Step 4. Turn the smoldot client into a raw client.
+    let raw_light_client = RawLightClient::builder()
+        .add_chain(polkadot_chain_id, polkadot_json_rpc_responses)
```

Comment:
hmm, so this a little complicated to first "add the chain" and then after call "for_chain".

I don't follow why adding the chain isn't sufficient AFAIU the `polkadot_json_rpc_responses` is just a stream that needs to polled or spawned in background task

Reply:
The `rpc_stream` needs to be provided while spawning the light-client at the moment.

It might be possible to introduce another subxt frontend -> subxt backend to propagate this to the smoldot client from the background task and add the `rpc_stream` to the streams polled in the backend. That seemed a bit more complicated than this, would love to hear your opinion on this 🙏

Reply:
It's fine for now but perhaps worth writing up an issue to look into.

My point is initializing and using the API is a bit complicated but 
looks to mainly because expose some of smoldots APIs.

Reply:
Yep that's a good point, raised https://github.com/paritytech/subxt/issues/1253 to handle it later, thanks!